service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/any/flexibleresource/ControlledFlexibleResource.java

```diff
@@ -0,0 +1,364 @@
+  public ControlledFlexibleResource(
+      @JsonProperty("workspaceId") UUID workspaceId,
+      @JsonProperty("resourceId") UUID resourceId,
+      @JsonProperty("name") String name,
+      @JsonProperty("description") String description,
+      @JsonProperty("cloningInstructions") CloningInstructions cloningInstructions,
+      @JsonProperty("assignedUser") String assignedUser,
+      @JsonProperty("privateResourceState") PrivateResourceState privateResourceState,
+      @JsonProperty("accessScope") AccessScopeType accessScope,
+      @JsonProperty("managedBy") ManagedByType managedBy,
+      @JsonProperty("applicationId") String applicationId,
+      @JsonProperty("resourceLineage") List<ResourceLineageEntry> resourceLineage,
+      @JsonProperty("properties") Map<String, String> properties,
+      @JsonProperty("createdByEmail") String createdByEmail,
+      @JsonProperty("createdDate") OffsetDateTime createdDate,
+      @JsonProperty("lastUpdatedByEmail") String lastUpdatedByEmail,
+      @JsonProperty("lastUpdatedDate") OffsetDateTime lastUpdatedDate,
+      @JsonProperty("region") String region,
+      @JsonProperty("typeNamespace") String typeNamespace,
+      @JsonProperty("type") String type,
+      @JsonProperty("data") String data) {
+    super(
+        ControlledResourceFields.builder()
+            .workspaceUuid(workspaceId)
+            .resourceId(resourceId)
+            .name(name)
+            .description(description)
+            .cloningInstructions(cloningInstructions)
+            .assignedUser(assignedUser)
+            .accessScope(accessScope)
+            .managedBy(managedBy)
+            .applicationId(applicationId)
+            .privateResourceState(privateResourceState)
+            .resourceLineage(resourceLineage)
+            .properties(properties)
+            .createdByEmail(createdByEmail)
+            .createdDate(createdDate)
+            .lastUpdatedByEmail(lastUpdatedByEmail)
+            .lastUpdatedDate(lastUpdatedDate)
+            .region(region)
+            .build());
+    this.typeNamespace = typeNamespace;
+    this.type = type;
+    this.data = data;
+    validate();
+  }
+
+  /*
+     // TODO: PF-2512 remove constructor above and enable this constructor
+   @JsonCreator
+   public FlexibleResourceResource(
+       @JsonProperty("wsmResourceFields") WsmResourceFields resourceFields,
+       @JsonProperty("wsmControlledResourceFields")
+           WsmControlledResourceFields controlledResourceFields,
+       @JsonProperty("type") String type,
+       @JsonProperty("typeNamespace") String typeNamespace,
+       @JsonProperty("data") Object data)
+  {
+     super(resourceFields, controlledResourceFields);
+     this.typeNamespace = typeNamespace;
+     this.type = type;
+     this.data = data;
+     validate();
+   }
+    */
+
+  // Constructor for the builder
+  private ControlledFlexibleResource(
+      ControlledResourceFields common, String typeNamespace, String type, String data) {
```

Comment:
`@Nullable`  here also, I think

I think it is worth marking it throughout the code. It is needed in a couple of other places. Helps us catch spots where we are not handling null.